### DISCUSSION TITLE
PyInstaller cannot correctly package the plugin library ruamel.yaml.string of ruamel.yaml

### DISCUSSION THREAD

**Discussion by kaixinol** [pyinstaller/pyinstaller#8960] [Help] [answered]
I'm trying to package one of my projects [GameYamlSpiderAndGenerator](https://github.com/FurryGamesIndex/GameYamlSpiderAndGenerator/), I created a py file in the upper directory for packaging
```python
import gameyamlspiderandgenerator.__main__
```
But apparently pyinstaller can't find this library
```log
Traceback (most recent call last):
File "pkg.py", line 2, in <module>
File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
File "PyInstaller/loader/pyimod02_importers.py", line 384, in exec_module
File "gameyamlspiderandgenerator/__main__.py", line 96, in <module>
File "gameyamlspiderandgenerator/util/fgi_yaml.py", line 34, in __str__
File "fgi_yaml_formattor/__init__.py", line 32, in dump_to_yaml
AttributeError: 'YAML' object has no attribute 'dump_to_string'
[PYI-15044:ERROR] Failed to execute script 'pkg' due to unhandled exception!
```
So I added the `--hidden-import ruamel.yaml.string \` option
But the log shows that pyinstaller still doesn't find it
full command line:
```bash
PYTHONPATH=./GameYamlSpiderAndGenerator/.venv/lib/python3.12/site-packages pyinstaller --onefile \
            --name GameYamlSpider \
            --add-data "gameyamlspiderandgenerator/plugin:gameyamlspiderandgenerator/plugin" \
            --add-data "./GameYamlSpiderAndGenerator/.venv/lib/python3.12/site-packages/language_data/data:language_data/data" \
            --hidden-import gameyamlspiderandgenerator.plugin.gcores \
            --hidden-import gameyamlspiderandgenerator.plugin.itchio \
            --hidden-import gameyamlspiderandgenerator.plugin.steam \
            --hidden-import yamlgenerator_hook_search \
            --hidden-import language_data \
            --hidden-import ruamel.yaml.string \
            --hidden-import yamlgenerator_hook_validate \
            pkg.py
```

**Comment by bwoodsend** [comment]
Sounds more likely to be environment mixup rather than anything to do with PyInstaller and adding a venv's site-packages to `PYTHONPATH` is a pretty good way to achieve that. Why don't you try:
```
./GameYamlSpiderAndGenerator/.venv/bin/pip install pyinstaller
./GameYamlSpiderAndGenerator/.venv/bin/pyinstaller --onefile \
   ... same options as before ... \
   pkg.py
```

**Comment by rokm** [comment, marked as answer]
Looks like `ruamel/yaml/string/__plug_in__.py` needs to be collected as a data file as well, based on [this plugin listing code from `ruamel.yaml`](https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/main.py#l757).

So `--add-data ./GameYamlSpiderAndGenerator/.venv/lib/python3.12/site-packages/ruamel/yaml/string/__plug_in__.py:ruamel/yaml/string`.

But I'll look into adding a hook so that this will be handled automatically and out-of-the-box (i.e., collect the submodules and force collection mode as pyz+py).

**Reply by rokm** [reply]
You can also try using the hook from https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/844.
